lib/Sema/SemaBounds.cpp

```diff
@@ -1418,20 +1434,280 @@ namespace {
+      }
+
+      ConstantSizedRange(Sema &S, const BoundsExpr *Base,
+                         llvm::APSInt &LowerOffset,
+                         llvm::APSInt &UpperOffset) :
+        S(S), Base(Base), LowerOffset(LowerOffset), UpperOffset(UpperOffset) {
+      }
 
-    // Check that SrcBounds implies that DeclaredBounds are provably
-    // true.
-    bool CheckBoundsDeclIsProvable(const BoundsExpr *DeclaredBounds,
-                                   const BoundsExpr *SrcBounds) {
+      // Is R in range of this range?
+      ProofResult InRange(ConstantSizedRange &R, ProofFailure &Cause) {
+        if (Lexicographic(S.Context, nullptr).CompareExpr(Base, R.Base) ==
+            Lexicographic::Result::Equal) {
+          ProofResult Result = ProofResult::True;
+          if (LowerOffset > R.LowerOffset) {
+            Cause = CombineFailures(Cause, ProofFailure::LowerBound);
+            Result = ProofResult::False;
+          }
+          if (UpperOffset < R.UpperOffset) {
+            Cause = CombineFailures(Cause, ProofFailure::UpperBound);
+            Result = ProofResult::False;
+          }
+          return Result;
+        }
+        return ProofResult::Maybe;
+      }
+
+      llvm::APSInt GetWidth() {
+        return UpperOffset - LowerOffset;
+      }
+
+      void SetBase(const Expr *B) {
+        Base = B;
+      }
+
+      void SetLower(llvm::APSInt &Lower) {
+        LowerOffset = Lower;
+      }
+
+      void SetUpper(llvm::APSInt &Upper) {
+        UpperOffset = Upper;
+      }
+
+      void Dump(raw_ostream &OS) {
+        OS << "ConstantRange:\n";
+        OS << "Base: ";
+        if (Base)
+          Base->dump(OS);
+        else
+          OS << "nullptr\n";
+        SmallString<12> Str1;
+        SmallString<12> Str2;
+        LowerOffset.toString(Str1);
+        UpperOffset.toString(Str2);
+        OS << "Lower offset:" << Str1 << "\nUpper offset:" << Str2 << "\n";
+      }
+    };
+
+    // Convert an expression E into a base and offset form.
+    // - If E is pointer arithmetic involving addition or subtraction of a
+    //   constant integer, return the base and offset.
+    // - Otherwise set base to E and offset to 0.
+    // TODO: we use signed integers to represent the result of the Offset.
+    // We can't represent unsigned offsets larger the the maximum signed
+    // integer that will fit pointer width.
+    void SplitIntoBaseAndOffset(const Expr *E, unsigned PointerWidth,
+                                const Expr *&Base, llvm::APSInt &Offset) {
+      if (const BinaryOperator *BO = dyn_cast<BinaryOperator>(E->IgnoreParens())) {
+        if (BO->isAdditiveOp()) {
+          Expr *Other = nullptr;
+          if (BO->getLHS()->getType()->isPointerType()) {
+            Base = BO->getLHS();
+            Other = BO->getRHS();
+          } else if (BO->getRHS()->getType()->isPointerType()) {
+            Base = BO->getRHS();
+            Other = BO->getLHS();
+          } else
+            goto exit;
```

Comment:
It seems this goes to `exit` if neither the LHS or RHS have pointer type. Except you also goto `exit` in error conditions. It's not clear these should have the same logic.

Reply:
I'll add a comment explain what's going on here.  We try to split an expression into a base + offset, where offset is in characters.  If we can't, for whatever reason, we return (E, 0).   

This includes pointer arithmetic overflow (where the size of the object times a constant integer overflows).  Eventually we ought to flag that as another error, but right now we're being imprecise for large unsigned integer offsets, so that can wait for another day.